compiler/lib/ir.mli

```diff
@@ -0,0 +1,173 @@
+(* Js_of_ocaml compiler
+ * http://www.ocsigen.org/js_of_ocaml/
+ * Copyright (C) 2010 Jérôme Vouillon
+ * Laboratoire PPS - CNRS Université Paris Diderot
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, with linking exception;
+ * either version 2.1 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+
+(*
+ * IR: A constrained intermediate language to model various language backends.
+ *
+ * IR has semantics similar to a very narrow subset of JavaScript, with the
+ * addition of special nodes that track types, externs, and different shapes of
+ * structures emitted by the ocaml compiler.
+ *
+ *)
+
+open Stdlib
+
+type array_literal = element_list
+
+and element_list = expression option list
+
+and binop =
+  | Eq
+  | StarEq
+  | SlashEq
+  | ModEq
+  | PlusEq
+  | MinusEq
+  | Or
+  | And
+  | Bor
+  | Bxor
+  | Band
+  | EqEq
+  | NotEq
+  | FloatEqEq
+  | FloatNotEq
+  | EqEqEq
+  | NotEqEq
+  | Lt
+  | Le
+  | Gt
+  | Ge
+  | FloatLt
+  | FloatLe
+  | FloatGt
+  | FloatGe
+  | InstanceOf
+  | Lsl
+  | Lsr
+  | Asr
+  | Plus
+  | FloatPlus
+  | Minus
+  | FloatMinus
+  | Mul
+  | FloatMul
+  | Div
+  | FloatDiv
+  | Mod
+  | FloatMod
+
+and unop =
+  | Not
+  | Neg
+  | FloatNeg
+  | Typeof
+  | IsInt
+  | ToInt
+  | ToBool
+  | IntToString
+  | FloatToInt
+  | IntToFloat
+  | Void
+  | Delete
+  | Bnot
+
+and arguments = expression list
+
+and property_name_and_value_list = (Id.property_name * expression) list
+
+and raw_segment =
+  | RawText of string
+  | RawSubstitution of expression
+
+and expression =
+  | ERaw of raw_segment list
```

Comment:
Can we leave `ERaw` to a later PR

Reply:
I've simplified `ERaw` as much as possible, but I'm not sure how to avoid it when plugging the new IR. `caml_js_expr` is currently parsing the JS code into the tree and embeds that directly. Converting JS tree to IR does not sound like a great idea.